core/new-gui/src/app/common/service/user/user-dictionary/dictionary.service.ts

```diff
@@ -0,0 +1,518 @@
+   * and the proxy dictionary only changes if those requests are successful.
+   * *Warning: proxy dictionaries are synced w/ backend each getAll() and should be correct 99.99% of the time.
+   * However they may fall out of sync due to unforseen networking errors.*
+   * @returns Proxy of a local copy once initial getAll() is finished
+   */
+  public getUserDictionaryAsync(): Observable<UserDictionary> {
+    return from(this.ready.promise.then(() => this.getUserDictionary()));
+  }
+
+  /**
+   * initializes proxy dictionaries by calling getAll() and then setting dictionaryService.ready state
+   */
+  private initLocalDict() {
+    let resolveReady: (read: boolean) => void;
+
+    this.ready = {promise: new Promise((resolver) => resolveReady = resolver), value: false};
+
+    // getAll automatically creates a dictionaryEvent that updates the localdict with the remote
+    this.getAll().subscribe({
+      next: () => {
+        this.ready.value = true;
+        resolveReady(true);
+      },
+      error: (error: Error) => {
+        // user not logged in - print error but continue
+        if (error.message === 'Invalid session') {
+          console.error(error);
+        } else {
+          throw error;
+        }
+      }
+    });
+
+  }
+
+  /**
+   * subscribes to dictionaryEventStream and updates proxy dictionaries with each dictionary event
+   * warnings are printed when the local copy is found to be different from the backend
+   */
+  private handleDictionaryEventStream() {
+    this.dictionaryEventStream().subscribe(event => {
+      switch (event.type) {
+        case EVENT_TYPE.GET:
+          if (event.key in this.localUserDictionary &&
+            this.localUserDictionary[event.key] !== event.value) {
+            console.warn(`[user-dictionary service] Dictionary desynchronized at key "${event.key}": `
+              + `locally had ${this.localUserDictionary[event.key]} but remote reported  ${event.value}`);
+          }
+          this.localUserDictionary[event.key] = event.value;
+          break;
+
+        case EVENT_TYPE.SET:
+          this.localUserDictionary[event.key] = event.value;
+          break;
+
+        case EVENT_TYPE.DELETE:
+          delete this.localUserDictionary[event.key];
+          break;
+
+        case EVENT_TYPE.GET_ALL:
+          if (!isEqual(this.localUserDictionary, event.value) && !isEqual(this.localUserDictionary, {})) {
+            console.warn(`[user-dictionary service] Dictionary was desynchronized, local had `
+              + `${this.localUserDictionary}, but remote reported ${event.value}`);
+          }
+          // setting this.localUserDictionary = event.value would not
+          // change the references to this.localUserDictionary used in all proxy dictionaries
+          for (const key in this.localUserDictionary) {
+            delete this.localUserDictionary[key];
+          }
+          Object.assign(this.localUserDictionary, event.value);
```

Comment:
I understand this part is about CUID operations with the local dictionary copy. It will be good to define methods to make it more readable. The handler of `GET` and `GET_ALL` seem to be mergeable (with methods).